File: LeetCode/152. Maximum Product Subarray.java
/***  KNOWLEDGE IS POWER  ***/

class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length, res = nums[0], l = 1, r = 1;
        for (int i = 0; i < n; i++) {
            l =  (l) * nums[i];
            r =  (r) * nums[n - 1 - i];
            res = Math.max(res,Math.max(l,r));
            if(l == 0)
                l = 1 ;
            if(r == 0)
                r = 1 ;
        }
        return res;
    }
}

File: LeetCode/152. Maximum Product Subarray.py
### KNOWLEDGE IS POWER  ###

class Solution(object):
    def maxProduct(self, nums):
        curMin, curMax, ans = nums[0], nums[0], nums[0]
        for i in range(1, len(nums)):
            tmp = curMax
            curMax = max(nums[i], nums[i]*tmp, nums[i]*curMin)
            curMin = min(nums[i], nums[i]*tmp, nums[i]*curMin)
            if curMax > ans:
                ans = curMax
        return ans
